frontend/src/scratch/Scratch.tsx

```diff
@@ -184,6 +184,18 @@ export default function Scratch() {
         debouncedCompile()
     }, [debouncedCompile, compiler?.compiler, compiler?.cc_opts])
 
+    const formatDiffText = (texts) => {
+        return texts.map((t) => {
+            if (t.format == "rotation") {
+                return <span class={styles[`diff-rotation-${t.index % 9}`]}>{t.text}</span>
+            } else if (t.format) {
+                return <span class={styles[`diff-${t.format}`]}>{t.text}</span>
+            } else {
+                return <span>{t.text}</span>
+            }
+        })
+    }
```

Comment:
This function can be raised into module-level scope instead of placing it inside `Scratch` IMO